### PR TITLE
Extend player profile API to support skin changes

### PR DESCRIPTION
**Pull request by NickAcPT**
Added code that refreshes the player's skin by sending packets with a special order, telling the client to respawn the player and re-apply the game profile

### CODE REVIEW

File: Spigot-Server-Patches/0336-Extend-player-profile-API-to-support-skin-changes.patch [hunk 1]
```diff
@@ -0,0 +1,185 @@
...
++                handle.playerConnection.sendPacket(legs);
++                handle.playerConnection.sendPacket(feet);
++
++            } else {
++                //Just send player update
++                ((CraftPlayer) pOnline).reregisterPlayer(this.getHandle());
++            }
++        }
++
++        }
```
Comment (mikroskeem) [inline Spigot-Server-Patches/0336-Extend-player-profile-API-to-support-skin-changes.patch]: Indent is messed up here

File: Spigot-Server-Patches/0336-Extend-player-profile-API-to-support-skin-changes.patch [hunk 2]
```diff
@@ -0,0 +1,185 @@
...
+-            player.getBukkitEntity().reregisterPlayer(self);
+-        }
++        refreshPlayer();
+     }
+     public PlayerProfile getPlayerProfile() {
+         return new CraftPlayerProfile(this).clone();
+     }
++
++    private void refreshPlayer() {
++        //Get the player's dimension
```
Comment (mikroskeem) [inline Spigot-Server-Patches/0336-Extend-player-profile-API-to-support-skin-changes.patch]: In my opinion whitespace after `//` would be nice, e.g `// Get the player's dimension`

File: Spigot-Server-Patches/0336-Extend-player-profile-API-to-support-skin-changes.patch [hunk 3]
```diff
@@ -0,0 +1,185 @@
...
++        int dimension = getWorld().getEnvironment().getId();
++
++        //Respawn packet
```
Comment (mikroskeem) [inline Spigot-Server-Patches/0336-Extend-player-profile-API-to-support-skin-changes.patch]: You can remove these comments, they are redundant.

**Comment by aikar** [comment]
Oh blah, I didn't notice you reopened new PR. Please see review feedback on last PR: https://github.com/PaperMC/Paper/pull/1243#pullrequestreview-139257082

File: Spigot-Server-Patches/0336-Extend-player-profile-API-to-support-skin-changes.patch [hunk 4]
```diff
@@ -0,0 +1,177 @@
...
+     private static void copyProfileProperties(GameProfile source, GameProfile target) {
+diff --git a/src/main/java/net/minecraft/server/WorldServer.java b/src/main/java/net/minecraft/server/WorldServer.java
+index dd78a87b..e3f7be28 100644
+--- a/src/main/java/net/minecraft/server/WorldServer.java
++++ b/src/main/java/net/minecraft/server/WorldServer.java
+@@ -135,6 +135,12 @@ public class WorldServer extends World implements IAsyncTaskHandler {
+         return this;
+     }
+ 
++    // Paper start - Provide dimension number of world
```
Comment (aikar) [inline Spigot-Server-Patches/0336-Extend-player-profile-API-to-support-skin-changes.patch]: I rather us just make the field public instead, more consistent in style.

File: Spigot-Server-Patches/0336-Extend-player-profile-API-to-support-skin-changes.patch [hunk 5]
```diff
@@ -0,0 +1,177 @@
...
+ import org.bukkit.craftbukkit.advancement.CraftAdvancementProgress;
++import org.bukkit.craftbukkit.inventory.CraftItemStack;
+ import org.bukkit.craftbukkit.map.CraftMapView;
+ import org.bukkit.craftbukkit.map.RenderData;
+ import org.bukkit.craftbukkit.scoreboard.CraftScoreboard;
+@@ -1172,14 +1173,102 @@ public class CraftPlayer extends CraftHumanEntity implements Player {
+     public void setPlayerProfile(PlayerProfile profile) {
+         EntityPlayer self = getHandle();
+         self.setProfile(CraftPlayerProfile.asAuthlibCopy(profile));
+-        List<EntityPlayer> players = server.getServer().getPlayerList().players;
```
Comment (aikar) [inline Spigot-Server-Patches/0336-Extend-player-profile-API-to-support-skin-changes.patch]: As I said on last PR, this PR should only need to scope itself to fixing the player seeing its own skin.

Leave this change as it was, as it was fine for OTHER Entities. 
All the logic you wrote below for other entities is re-implementing the entire Entity Tracking system which we do not want to do.

So gut all that, keep this for statement, and make refreshPlayer() focused only on it self.

File: Spigot-Server-Patches/0336-Extend-player-profile-API-to-support-skin-changes.patch [hunk 6]
```diff
@@ -0,0 +1,177 @@
...
++
++    private void refreshPlayer() {
+
++        int dimension = getWorld().getEnvironment().getId();
++
++        Packet respawn = new PacketPlayOutRespawn(dimension, ((WorldServer)getHandle().getWorld()).worldData.getDifficulty(), ((WorldServer)getHandle().getWorld()).worldData.getType(), EnumGamemode.getById(getHandle().playerInteractManager.getGameMode().getId()));
++
++        Location l = getLocation();
++
++        Packet pos = new PacketPlayOutPosition(l.getX(), l.getY(), l.getZ(), l.getYaw(), l.getPitch(), new HashSet<>(), 0);
```
Comment (aikar) [inline Spigot-Server-Patches/0336-Extend-player-profile-API-to-support-skin-changes.patch]: As said on the other review, gut all this, keep the reregisterPlayer loop above

File: Spigot-Server-Patches/0336-Extend-player-profile-API-to-support-skin-changes.patch [hunk 7]
```diff
@@ -0,0 +1,177 @@
...
++
++                //Send armor
++                handle.playerConnection.sendPacket(helm);
++                handle.playerConnection.sendPacket(chest);
++                handle.playerConnection.sendPacket(legs);
++                handle.playerConnection.sendPacket(feet);
++
++            } else {
++                //Just send player update
++                ((CraftPlayer) pOnline).reregisterPlayer(this.getHandle());
```
Comment (aikar) [inline Spigot-Server-Patches/0336-Extend-player-profile-API-to-support-skin-changes.patch]: This was actually a major issue. We don't want to touch anything for the player if the the receiving player cant see the source player

Comment (aikar) [reply Spigot-Server-Patches/0336-Extend-player-profile-API-to-support-skin-changes.patch]: pOnline.canSee(this)

**Review by aikar — CHANGES_REQUESTED** [review]

File: Spigot-Server-Patches/0336-Extend-player-profile-API-to-support-skin-changes.patch [hunk 8]
```diff
@@ -0,0 +1,81 @@
...
+diff --git a/src/main/java/org/bukkit/craftbukkit/CraftWorld.java b/src/main/java/org/bukkit/craftbukkit/CraftWorld.java
+index 210e3bc4..b0335684 100644
+--- a/src/main/java/org/bukkit/craftbukkit/CraftWorld.java
++++ b/src/main/java/org/bukkit/craftbukkit/CraftWorld.java
+@@ -101,6 +101,12 @@ public class CraftWorld implements World {
+     }
+     // Paper end
+ 
++    // Paper start - Provide dimension number of world
++    public int getDimension() {
```
Comment (stephan-gh) [inline Spigot-Server-Patches/0336-Extend-player-profile-API-to-support-skin-changes.patch]: This looks unused now.

**Review by stephan-gh — CHANGES_REQUESTED** [review]

File: Spigot-Server-Patches/0336-Extend-player-profile-API-to-support-skin-changes.patch (hunk 8 above)
Comment (NickAcPT) [reply Spigot-Server-Patches/0336-Extend-player-profile-API-to-support-skin-changes.patch]: oh. I forgot to delete that. Just a sec.

File: Spigot-Server-Patches/0336-Extend-player-profile-API-to-support-skin-changes.patch [hunk 9]
```diff
@@ -0,0 +1,64 @@
...
+     public PlayerProfile getPlayerProfile() {
+         return new CraftPlayerProfile(this).clone();
+     }
++
++    private void refreshPlayer() {
++        EntityPlayer entityplayer = getHandle();
++
++        Packet respawn = new PacketPlayOutRespawn(entityplayer.dimension, entityplayer.world.getDifficulty(), entityplayer.world.getWorldData().getType(), entityplayer.playerInteractManager.getGameMode());
++        Location l = getLocation();
++        Packet pos = new PacketPlayOutPosition(l.getX(), l.getY(), l.getZ(), l.getYaw(), l.getPitch(), new HashSet<>(), 0);
```
Comment (aikar) [inline Spigot-Server-Patches/0336-Extend-player-profile-API-to-support-skin-changes.patch]: Ok going to bunch everything up into one review comment, overall this is looking a lot better now.

But lets clean up some naming/impl details.

1) remove the import up top for CraftItemStack, remnant from previous version
2) rename l to loc, we prefer to not have 1 letter variables outside of positional stuff like x/y/z
3) rename playerCon to just connection
4) inline the packets so its connection.sendPacket(new PacketPlayOut*), we don't need variables for that too.
5) Add a comment on where this logic can be found in EntityPlayer or such incase future updates change the logic and we need to update this.

I think that'll cover it on quality wise and be good to go at that point.

**Review by aikar — CHANGES_REQUESTED** [review]

**Pull request merged**